### PR TITLE
Opt into unsafe lifecycle warnings without async tree

### PR DESCRIPTION
**Pull request by bvaughn**
Opts a subtree into the same warnings as `AsyncComponent` does, but does not actually enable async rendering. This same mechanism could be extended for similar warnings in the future (eg compiler, React Fabric).

Following precedent, this component is exposed via `React.StrictMode`. It is used like a `Fragment` to wrap an element:

```js
const {StrictMode} = React;

<StrictMode>
  {/* Subtree here */}
</StrictMode>
```

This branch builds on top of #12060.

### CODE REVIEW

File: packages/react/src/ReactBaseClasses.js [hunk 1]
```diff
@@ -117,44 +117,56 @@ if (__DEV__) {
...
   this.updater = updater || ReactNoopUpdateQueue;
 }
+configurePrototype(PureComponent, {isPureReactComponent: true});
 
-function ComponentDummy() {}
-ComponentDummy.prototype = Component.prototype;
-const pureComponentPrototype = (PureComponent.prototype = new ComponentDummy());
-pureComponentPrototype.constructor = PureComponent;
-// Avoid an extra prototype jump for these methods.
-Object.assign(pureComponentPrototype, Component.prototype);
```
Comment (sebmarkbage) [inline packages/react/src/ReactBaseClasses.js]: Are we sure this matters these days in any significant way? Maybe we can just kill this.

**Comment by sebmarkbage** [comment]
@acdlite Originally we thought that we needed to be able to inherit AsyncComponent instead of wrapping, because at the top level it is tricky since it affects the return value of render. However, if you're upgrading at the top level, you probably want to use the createRoot API instead. Do we still think that we want the inheritance mechanism or should we switch to what context and fragment do?

**Comment by bvaughn** [comment]
> Duplication is not a good enough reason to introduce an abstraction.

The code felt crufty and repetitive before, so I proposed a change. I'll revert it.

> Maybe we can rename it to something like `<EnableFutureCompatibilityChecks />` or something? I don't know.

Interesting. I like this suggestion. Not sure about _that name_ but I like the suggestion.

**Comment by gaearon** [comment]
To give some background on the abstraction thing, the code feeling crufty/duplicated is how we ended up with a lot of things in the Stack implementation (like "Transactions") that ended up being hard to work with. So these days we try to err on the side of making *removal* of the code easy rather than making it feel dry.

You know why the abstraction was added and thus know when to remove it. But other people often don’t, and as time goes everybody is afraid to rip out each other’s abstractions, and we end up with overabstracted code that *seems* to be necessary but no one’s really sure. Hope this makes sense 😛

**Comment by bvaughn** [comment]
Sure, it makes sense. I didn't feel strongly about the change anyway. I mentioned in the initial PR description that I would be happy to revert it if there were any concerns, for this reason ^

**Comment by bvaughn** [comment]
Based on discussion above and in person (during lunch) I've updated this PR as follows:
* The new component type is named `StrictMode` (rather than anything to do with async).
* The new internal context flag is also named `StrictMode`.
* Async subtrees are also `StrictMode` subtrees and all of the previous async warning checks are now based on this flag instead.
* The `ReactDebugAsyncWarnings` helper has been renamed `ReactStrictModeWarnings` to avoid confusion.
* A warning message has been updated with a new fb.me link, https://fb.me/react-strict-mode-warnings

**Comment by sebmarkbage** [comment]
What did you think of @tomocchino's `ReactDOM.renderStrict(...)` proposal to opt in a whole tree?

That would eliminate the need for base class based opt in `class Foo extends StrictMode` which seems weird.

**Comment by bvaughn** [comment]
> What did you think of @tomocchino's `ReactDOM.renderStrict(...)` proposal to opt in a whole tree?

I think I like it as a follow-up option, but it doesn't seem sufficient to replace `React.StrictComponent`.

**Comment by sebmarkbage** [comment]
If you're testing your children you can always wrap `return <StrictMode>{myChildren}</StrictMode>`. The only thing it helps is with testing yourself but I'm not sure how valuable that is.

**Comment by bvaughn** [comment]
Oh I think I misunderstood your question!

**Comment by bvaughn** [comment]
Okay! I converted `StrictMode` to a symbol, verified all previous tests passed, and added a couple of new tests to cover going between `StrictMode` -> `Fragment`, `Fragment` -> `StrictMode`, and `StrictMode` -> `StrictMode`.

File: packages/shared/ReactSymbols.js [hunk 2]
```diff
@@ -24,6 +24,9 @@ export const REACT_PORTAL_TYPE = hasSymbol
 export const REACT_FRAGMENT_TYPE = hasSymbol
   ? Symbol.for('react.fragment')
   : 0xeacb;
+export const REACT_USE_STRICT_TYPE = hasSymbol
+  ? Symbol.for('react.use_strict')
```
Comment (sebmarkbage) [inline packages/shared/ReactSymbols.js]: Let's call it the same thing as what we export. `react.strict_mode`

`use strict` is JavaScript.

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 3]
```diff
@@ -639,20 +639,23 @@ export default function(
...
+    if (workInProgress.type != null && workInProgress.type.prototype != null) {
+      const prototype = workInProgress.type.prototype;
+
+      if (enableAsyncSubtreeAPI) {
+        if (prototype.unstable_isAsyncReactComponent === true) {
+          workInProgress.internalContextTag |= AsyncUpdates;
+        }
+      }
+
+      if (prototype.__reactStrictMode === true) {
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: This seems like it is not necessary anymore.

File: packages/react/src/ReactBaseClasses.js [hunk 4]
```diff
@@ -153,8 +151,30 @@ asyncComponentPrototype.constructor = AsyncComponent;
...
   return this.props.children;
 };
 
-export {Component, PureComponent, AsyncComponent};
+/**
+ * Special component type that enables forward-looking, advanced warnings.
+ * For examples, it detects async-unsafe lifecycles (without actually enabling async mode).
+ * It can also detect render-phase side effects by double-invoking certain methods.
+ */
+function StrictMode(props, context, updater) {
```
Comment (sebmarkbage) [inline packages/react/src/ReactBaseClasses.js]: We can remove this now.

**Comment by bvaughn** [comment]
Whoops. I made a silly oversight. Give me a moment.

File: packages/react/src/ReactBaseClasses.js (hunk 4 above)
Comment (bvaughn) [reply packages/react/src/ReactBaseClasses.js]: Yeah, this was the silly oversight ;)

File: packages/shared/ReactSymbols.js (hunk 2 above)
Comment (gaearon) [reply packages/shared/ReactSymbols.js]: But ... React is JavaScript! 😄

**Comment by gaearon** [comment]
I *love* the strict mode parallel. This is a much easier story to convey.

**Comment by bvaughn** [comment]
Okay. A new type-of-work `Mode` has been added.

**Pull request merged**